File: frontend/src/types/waitEvent.ts
/**
 * TypeScript type definitions for wait events.
 * These types mirror the backend Pydantic schemas.
 */

/**
 * Individual wait event from ASH.
 */
export interface WaitEvent {
  sql_id?: string;
  sample_time?: string;
  session_id?: number;
  session_serial?: number;
  event?: string;
  wait_class?: string;
  wait_time?: number;
  time_waited?: number;
  p1text?: string;
  p1?: number;
  p2text?: string;
  p2?: number;
  p3text?: string;
  p3?: number;
  current_obj?: number;
  current_file?: number;
  current_block?: number;
  session_state?: string;
  blocking_session?: number;
  blocking_session_serial?: number;
  sql_plan_hash_value?: number;
  sql_plan_line_id?: number;
  category?: string;
}

/**
 * Current wait event from V$SESSION_WAIT.
 */
export interface CurrentWaitEvent {
  sid: number;
  serial: number;
  username?: string;
  program?: string;
  sql_id?: string;
  event: string;
  wait_class: string;
  state: string;
  wait_time?: number;
  seconds_in_wait?: number;
  p1text?: string;
  p1?: number;
  p2text?: string;
  p2?: number;
  p3text?: string;
  p3?: number;
  category?: string;
}

/**
 * Aggregated wait event summary.
 */
export interface WaitEventSummary {
  event: string;
  wait_class: string;
  category: string;
  wait_count: number;
  total_time_waited: number;
  avg_time_waited?: number;
  max_time_waited?: number;
  percentage: number;
}

/**
 * Wait event category aggregation.
 */
export interface WaitEventCategory {
  category: string;
  name: string;
  color: string;
  total_time_waited: number;
  wait_count?: number;
  event_types?: number;
  session_count?: number;
  total_wait_time?: number;
  percentage?: number;
}

/**
 * Time bucket for wait event timeline.
 */
export interface WaitEventTimelineBucket {
  sample_minute: string;
  wait_class: string;
  category: string;
  sample_count: number;
  total_time_waited: number;
}

/**
 * Blocking session information.
 */
export interface BlockingSession {
  blocking_session: number;
  blocking_session_serial: number;
  blocking_username?: string;
  blocking_program?: string;
  blocking_sql_id?: string;
  event: string;
  block_count: number;
}

/**
 * Wait event tuning recommendation.
 */
export interface WaitEventRecommendation {
  type: string;
  priority: 'low' | 'medium' | 'high';
  message: string;
  suggestions: string[];
  blocking_sessions?: BlockingSession[];
}

/**
 * Wait events data for a SQL_ID.
 */
export interface WaitEventsData {
  sql_id: string;
  hours_back: number;
  total_time_waited: number;
  total_samples: number;
  wait_events: WaitEvent[];
  category_breakdown: WaitEventCategory[];
  top_events: WaitEventSummary[];
  timeline: WaitEventTimelineBucket[];
  blocking_sessions: BlockingSession[];
  recommendations: WaitEventRecommendation[];
}

/**
 * API response for wait events by SQL_ID.
 */
export interface WaitEventsResponse {
  data: WaitEventsData;
}

/**
 * Wait events grouped by SQL_ID.
 */
export interface WaitEventSQLBreakdown {
  sql_id: string;
  session_count: number;
  events: string[];
}

/**
 * Current system wait events data.
 */
export interface CurrentWaitEventsData {
  timestamp: string;
  total_sessions: number;
  wait_events: CurrentWaitEvent[];
  category_breakdown: WaitEventCategory[];
  sql_id_breakdown: WaitEventSQLBreakdown[];
}

/**
 * API response for current system wait events.
 */
export interface CurrentWaitEventsResponse {
  data: CurrentWaitEventsData;
}

/**
 * Chart data for wait event visualization.
 */
export interface WaitEventChartData {
  name: string;
  value: number;
  percentage: number;
  color: string;
  category: string;
}

/**
 * Timeline chart data point.
 */
export interface TimelineDataPoint {
  time: string;
  [key: string]: number | string; // Dynamic keys for each category
}

File: backend/app/schemas/wait_event.py
"""Pydantic schemas for wait events."""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class WaitEvent(BaseModel):
    """Individual wait event from ASH or V$SESSION_WAIT."""

    sql_id: Optional[str] = Field(None, description="SQL ID")
    sample_time: Optional[str] = Field(None, description="Sample time")
    session_id: Optional[int] = Field(None, description="Session ID")
    session_serial: Optional[int] = Field(None, description="Session serial number")
    event: Optional[str] = Field(None, description="Wait event name")
    wait_class: Optional[str] = Field(None, description="Wait class")
    wait_time: Optional[int] = Field(None, description="Wait time in centiseconds")
    time_waited: Optional[int] = Field(None, description="Time waited in microseconds")
    p1text: Optional[str] = Field(None, description="Parameter 1 name")
    p1: Optional[int] = Field(None, description="Parameter 1 value")
    p2text: Optional[str] = Field(None, description="Parameter 2 name")
    p2: Optional[int] = Field(None, description="Parameter 2 value")
    p3text: Optional[str] = Field(None, description="Parameter 3 name")
    p3: Optional[int] = Field(None, description="Parameter 3 value")
    current_obj: Optional[int] = Field(None, description="Current object number")
    current_file: Optional[int] = Field(None, description="Current file number")
    current_block: Optional[int] = Field(None, description="Current block number")
    session_state: Optional[str] = Field(None, description="Session state")
    blocking_session: Optional[int] = Field(None, description="Blocking session ID")
    blocking_session_serial: Optional[int] = Field(
        None, description="Blocking session serial"
    )
    sql_plan_hash_value: Optional[int] = Field(None, description="SQL plan hash value")
    sql_plan_line_id: Optional[int] = Field(None, description="SQL plan line ID")
    category: Optional[str] = Field(None, description="Wait event category")

    class Config:
        from_attributes = True


class CurrentWaitEvent(BaseModel):
    """Current wait event from V$SESSION_WAIT."""

    sid: int = Field(..., description="Session ID")
    serial: int = Field(..., description="Session serial number")
    username: Optional[str] = Field(None, description="Username")
    program: Optional[str] = Field(None, description="Program name")
    sql_id: Optional[str] = Field(None, description="SQL ID")
    event: str = Field(..., description="Wait event name")
    wait_class: str = Field(..., description="Wait class")
    state: str = Field(..., description="Wait state")
    wait_time: Optional[int] = Field(None, description="Wait time")
    seconds_in_wait: Optional[int] = Field(None, description="Seconds in wait")
    p1text: Optional[str] = Field(None, description="Parameter 1 name")
    p1: Optional[int] = Field(None, description="Parameter 1 value")
    p2text: Optional[str] = Field(None, description="Parameter 2 name")
    p2: Optional[int] = Field(None, description="Parameter 2 value")
    p3text: Optional[str] = Field(None, description="Parameter 3 name")
    p3: Optional[int] = Field(None, description="Parameter 3 value")
    category: Optional[str] = Field(None, description="Wait event category")

    class Config:
        from_attributes = True


class WaitEventSummary(BaseModel):
    """Aggregated wait event summary."""

    event: str = Field(..., description="Wait event name")
    wait_class: str = Field(..., description="Wait class")
    category: str = Field(..., description="Wait event category")
    wait_count: int = Field(..., description="Number of waits")
    total_time_waited: float = Field(..., description="Total time waited (microseconds)")
    avg_time_waited: Optional[float] = Field(
        None, description="Average time waited (microseconds)"
    )
    max_time_waited: Optional[float] = Field(
        None, description="Maximum time waited (microseconds)"
    )
    percentage: float = Field(..., description="Percentage of total wait time")


class WaitEventCategory(BaseModel):
    """Wait event category aggregation."""

    category: str = Field(..., description="Category key")
    name: str = Field(..., description="Category display name")
    color: str = Field(..., description="Category color")
    total_time_waited: float = Field(..., description="Total time waited")
    wait_count: Optional[int] = Field(None, description="Number of waits")
    event_types: Optional[int] = Field(None, description="Number of event types")
    session_count: Optional[int] = Field(None, description="Number of sessions")
    total_wait_time: Optional[float] = Field(None, description="Total wait time (seconds)")
    percentage: Optional[float] = Field(None, description="Percentage of total time")


class WaitEventTimelineBucket(BaseModel):
    """Time bucket for wait event timeline."""

    sample_minute: str = Field(..., description="Sample time (minute)")
    wait_class: str = Field(..., description="Wait class")
    category: str = Field(..., description="Wait event category")
    sample_count: int = Field(..., description="Number of samples")
    total_time_waited: float = Field(..., description="Total time waited")


class BlockingSession(BaseModel):
    """Blocking session information."""

    blocking_session: int = Field(..., description="Blocking session ID")
    blocking_session_serial: int = Field(..., description="Blocking session serial")
    blocking_username: Optional[str] = Field(None, description="Blocking username")
    blocking_program: Optional[str] = Field(None, description="Blocking program")
    blocking_sql_id: Optional[str] = Field(None, description="Blocking SQL ID")
    event: str = Field(..., description="Wait event")
    block_count: int = Field(..., description="Number of times blocked")


class WaitEventRecommendation(BaseModel):
    """Wait event tuning recommendation."""

    type: str = Field(..., description="Recommendation type")
    priority: str = Field(..., description="Priority (low, medium, high)")
    message: str = Field(..., description="Recommendation message")
    suggestions: List[str] = Field(..., description="List of suggestions")
    blocking_sessions: Optional[List[BlockingSession]] = Field(
        None, description="Related blocking sessions"
    )


class WaitEventsResponse(BaseModel):
    """API response for wait events by SQL_ID."""

    data: Dict[str, Any] = Field(..., description="Wait event data")


class CurrentWaitEventsResponse(BaseModel):
    """API response for current system wait events."""

    data: Dict[str, Any] = Field(..., description="Current wait event data")


class WaitEventSQLBreakdown(BaseModel):
    """Wait events grouped by SQL_ID."""

    sql_id: str = Field(..., description="SQL ID")
    session_count: int = Field(..., description="Number of sessions")
    events: List[str] = Field(..., description="List of wait events")
